File: internal/api/sources.go
package api

import (
	"context"
	"errors"
	"net/http"
	"os"
	"strconv"

	"github.com/vesahyp/clavesa/internal/httputil"
	"github.com/vesahyp/clavesa/internal/preview"
)

// SourceRegistry is the service-layer interface the api package depends on
// for the ADR-017 source registry. Mirrors the relevant service.Service
// methods so internal/api stays free of an internal/service import.
type SourceRegistry interface {
	AddSource(spec SourceSpec) (SourceSpec, error)
	UpdateSource(name string, spec SourceSpec) (SourceSpec, error)
	ListSources() ([]SourceSpec, error)
	GetSource(name string) (SourceSpec, error)
	DeleteSource(name string, force bool) error
	AttachSource(dir, name, toNode, alias string) error
	PreviewRegistrySource(ctx context.Context, name string, offset, limit int) (*preview.PreviewResult, error)
}

// SourceSpec mirrors sources.Spec / service.SourceSpec at the api
// boundary. Field tags must stay byte-identical to the storage shape
// — the UI parses with the same names.
type SourceSpec struct {
	Name                      string   `json:"name"`
	Kind                      string   `json:"kind"`
	URL                       string   `json:"url,omitempty"`
	Bucket                    string   `json:"bucket,omitempty"`
	Prefix                    string   `json:"prefix,omitempty"`
	Format                    string   `json:"format,omitempty"`
	Credentials               string   `json:"credentials,omitempty"`
	Partitions                []string `json:"partitions,omitempty"`
	StartFrom                 string   `json:"start_from,omitempty"`
	ManageBucketNotifications bool     `json:"manage_bucket_notifications,omitempty"`
}

// SourceUsage names a pipeline that references the source. Returned in
// the 409 Conflict body of DELETE so the UI can render "used by N
// pipelines" inline.
type SourceUsage struct {
	PipelineDir string   `json:"pipeline_dir"`
	NodeIDs     []string `json:"node_ids"`
}

// SourcesHandler serves the /sources endpoints.
type SourcesHandler struct {
	svc SourceRegistry
}

// NewSourcesHandler returns a handler wired to a service-layer registry.
func NewSourcesHandler(svc SourceRegistry) *SourcesHandler {
	return &SourcesHandler{svc: svc}
}

// RegisterRoutes mounts CRUD + attach routes under the api mux.
func (h *SourcesHandler) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /sources", h.list)
	mux.HandleFunc("POST /sources", h.register)
	mux.HandleFunc("GET /sources/{name}", h.get)
	mux.HandleFunc("GET /sources/{name}/preview", h.preview)
	mux.HandleFunc("PUT /sources/{name}", h.update)
	mux.HandleFunc("DELETE /sources/{name}", h.delete)
	mux.HandleFunc("POST /sources/{name}/attach", h.attach)
}

type sourcesListResponse struct {
	Sources []SourceSpec `json:"sources"`
}

func (h *SourcesHandler) list(w http.ResponseWriter, r *http.Request) {
	list, err := h.svc.ListSources()
	if err != nil {
		httputil.WriteError(w, http.StatusInternalServerError, "list sources: "+err.Error())
		return
	}
	httputil.WriteJSON(w, http.StatusOK, sourcesListResponse{Sources: list})
}

func (h *SourcesHandler) get(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	spec, err := h.svc.GetSource(name)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			httputil.WriteError(w, http.StatusNotFound, "source not found: "+name)
			return
		}
		httputil.WriteError(w, http.StatusBadRequest, err.Error())
		return
	}
	httputil.WriteJSON(w, http.StatusOK, spec)
}

func (h *SourcesHandler) register(w http.ResponseWriter, r *http.Request) {
	req, ok := httputil.DecodeJSON[SourceSpec](w, r)
	if !ok {
		return
	}
	// Don't default Kind here — service.AddSource sniffs `s3://` URLs
	// and promotes kind=s3 with bucket/prefix derivation. Forcing
	// Kind="http" here would short-circuit that and 400 every UI POST
	// of an s3:// URL (the registerSource client doesn't know the
	// kind, only the URL).
	stored, err := h.svc.AddSource(req)
	if err != nil {
		httputil.WriteServiceError(w, err, http.StatusBadRequest)
		return
	}
	httputil.WriteJSON(w, http.StatusOK, stored)
}

// preview samples a registered source's raw data standalone — the HTTP
// twin of `clavesa source preview <name>`.
func (h *SourcesHandler) preview(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))
	limit := 50
	if v, err := strconv.Atoi(r.URL.Query().Get("limit")); err == nil && v > 0 {
		limit = v
	}
	res, err := h.svc.PreviewRegistrySource(r.Context(), name, offset, limit)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			httputil.WriteError(w, http.StatusNotFound, "source not found: "+name)
			return
		}
		httputil.WriteError(w, http.StatusBadRequest, err.Error())
		return
	}
	httputil.WriteJSON(w, http.StatusOK, res)
}

// update overwrites an existing source. The {name} path segment is the
// fixed registry key — a `name` in the body is ignored, since renaming
// is a delete + re-register, not an edit.
func (h *SourcesHandler) update(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	req, ok := httputil.DecodeJSON[SourceSpec](w, r)
	if !ok {
		return
	}
	stored, err := h.svc.UpdateSource(name, req)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			httputil.WriteError(w, http.StatusNotFound, "source not found: "+name)
			return
		}
		httputil.WriteError(w, http.StatusBadRequest, err.Error())
		return
	}
	httputil.WriteJSON(w, http.StatusOK, stored)
}

type deleteSourceResponse struct {
	Usages []SourceUsage `json:"usages,omitempty"`
}

// inUseConflicter lets the api layer pull the structured usage list out
// of a service-side ErrSourceInUse without importing internal/service.
type inUseConflicter interface {
	error
	InUseUsages() []SourceUsage
}

func (h *SourcesHandler) delete(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	force := r.URL.Query().Get("force") == "1"
	err := h.svc.DeleteSource(name, force)
	if err == nil {
		w.WriteHeader(http.StatusNoContent)
		return
	}
	if errors.Is(err, os.ErrNotExist) {
		httputil.WriteError(w, http.StatusNotFound, "source not found: "+name)
		return
	}
	// 409 + structured body when in use, so the UI can render the
	// dependency list inline.
	if c, ok := err.(inUseConflicter); ok {
		httputil.WriteJSON(w, http.StatusConflict, deleteSourceResponse{Usages: c.InUseUsages()})
		return
	}
	httputil.WriteError(w, http.StatusInternalServerError, err.Error())
}

type attachRequest struct {
	Dir   string `json:"dir"`
	To    string `json:"to"`
	Alias string `json:"alias"`
}

func (h *SourcesHandler) attach(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	req, ok := httputil.DecodeJSON[attachRequest](w, r)
	if !ok {
		return
	}
	if !httputil.RequireFields(w, map[string]string{"dir": req.Dir, "to": req.To}) {
		return
	}
	if err := h.svc.AttachSource(req.Dir, name, req.To, req.Alias); err != nil {
		httputil.WriteServiceError(w, err, http.StatusBadRequest)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

File: ui/src/lib/queries/sources.ts
/**
 * sources.ts — the workspace source + credential registries (ADR-017
 * slices 1 and 2): list hooks, register/update mutations, attach, and the
 * guarded deletes.
 */

import { useQuery } from "@tanstack/react-query";
import { z } from "zod";

import { request } from "@/api/client";
import { requestDelete, requestParsed, requestVoid } from "./core";

// ---------------------------------------------------------------------------
// Sources — workspace input source registry (ADR-017 slice 1)
// ---------------------------------------------------------------------------

const SourceSpec = z.object({
  name: z.string(),
  kind: z.string(),
  url: z.string().optional().default(""),
  bucket: z.string().optional().default(""),
  prefix: z.string().optional().default(""),
  format: z.string().optional().default(""),
  credentials: z.string().optional().default(""),
  partitions: z.array(z.string()).optional().default([]),
  start_from: z.string().optional().default(""),
  manage_bucket_notifications: z.boolean().optional().default(false),
});
export type SourceSpec = z.infer<typeof SourceSpec>;

// ---------------------------------------------------------------------------
// Credentials — workspace credentials registry (ADR-017 slice 2)
// ---------------------------------------------------------------------------

const CredentialSpec = z.object({
  name: z.string(),
  kind: z.string(),
  header_name: z.string().optional().default(""),
  value_prefix: z.string().optional().default(""),
  secret: z.string(),
  backend: z.string().optional().default(""),
});
export type CredentialSpec = z.infer<typeof CredentialSpec>;

const CredentialsListResponse = z.object({
  credentials: z.array(CredentialSpec).default([]),
});

/** GET /api/credentials — registered credentials in this workspace. */
export function useCredentials() {
  return useQuery({
    queryKey: ["credentials"],
    queryFn: async () => {
      const raw = await request<unknown>("/credentials");
      return CredentialsListResponse.parse(raw);
    },
  });
}

/** POST /api/credentials — register a new credential. */
export async function registerCredential(spec: {
  name: string;
  kind?: string;
  header_name?: string;
  value_prefix?: string;
  secret: string;
}): Promise<CredentialSpec> {
  return requestParsed("/credentials", CredentialSpec, {
    method: "POST",
    body: JSON.stringify({ kind: spec.kind ?? "header", ...spec }),
    errorLabel: "POST /credentials",
  });
}

/** DELETE /api/credentials/{name}. */
export async function deleteCredential(
  name: string,
  opts: { force?: boolean } = {},
): Promise<{ usages?: { source_name: string }[] } | null> {
  const params = opts.force ? "?force=1" : "";
  return requestDelete(
    `/credentials/${encodeURIComponent(name)}${params}`,
    `DELETE /credentials/${name}`,
  );
}

const SourcesListResponse = z.object({
  sources: z.array(SourceSpec).default([]),
});

/** GET /api/sources — registered sources in this workspace. */
export function useSources() {
  return useQuery({
    queryKey: ["sources"],
    queryFn: async () => {
      const raw = await request<unknown>("/sources");
      return SourcesListResponse.parse(raw);
    },
  });
}

/** POST /api/sources — register a new source. */
export async function registerSource(spec: {
  name: string;
  kind?: string;
  url?: string;
  bucket?: string;
  prefix?: string;
  format?: string;
  credentials?: string;
  partitions?: string[];
  start_from?: string;
  manage_bucket_notifications?: boolean;
}): Promise<SourceSpec> {
  // kind omitted on purpose — the server sniffs `s3://` vs `https://`
  // out of the URL field. Caller can still pass kind explicitly when
  // it knows.
  return requestParsed("/sources", SourceSpec, {
    method: "POST",
    body: JSON.stringify(spec),
    errorLabel: "POST /sources",
  });
}

/**
 * PUT /api/sources/{name} — overwrite an existing source's spec. The
 * name is the fixed registry key; renaming is a delete + re-register.
 */
export async function updateSource(
  name: string,
  spec: {
    name: string;
    kind?: string;
    url?: string;
    bucket?: string;
    prefix?: string;
    format?: string;
    credentials?: string;
    partitions?: string[];
    start_from?: string;
    manage_bucket_notifications?: boolean;
  },
): Promise<SourceSpec> {
  return requestParsed(`/sources/${encodeURIComponent(name)}`, SourceSpec, {
    method: "PUT",
    body: JSON.stringify(spec),
    errorLabel: `PUT /sources/${name}`,
  });
}

/**
 * POST /api/sources/{name}/attach — wire a registered source into a
 * transform's `inputs` map as `inputs[alias] = "sources.<name>"`. Returns
 * void (204 on success); body of the throw includes the server error
 * message on failure.
 */
export async function attachSource(
  name: string,
  body: { dir: string; to: string; alias?: string },
): Promise<void> {
  return requestVoid(`/sources/${encodeURIComponent(name)}/attach`, {
    method: "POST",
    body: JSON.stringify(body),
    errorLabel: `POST /sources/${name}/attach`,
    require204: true,
  });
}

/**
 * POST /api/pipeline/external-table/attach — wire a cross-pipeline or
 * external Glue table into a transform's `inputs` map as
 * `inputs[alias] = "<schema>.<table>"` (ADR-016 slice 2). HTTP twin of
 * the CLI `node connect --from-table` command. Returns the updated
 * graph on success.
 */
export async function attachExternalTable(body: {
  dir: string;
  ref: string;
  to: string;
  alias?: string;
}): Promise<void> {
  return requestVoid("/pipeline/external-table/attach", {
    method: "POST",
    body: JSON.stringify(body),
    errorLabel: "POST /pipeline/external-table/attach",
  });
}

/** DELETE /api/sources/{name}. Returns null on success, usage list on 409. */
export async function deleteSource(
  name: string,
  opts: { force?: boolean } = {},
): Promise<{ usages?: { pipeline_dir: string; node_ids: string[] }[] } | null> {
  const params = opts.force ? "?force=1" : "";
  return requestDelete(
    `/sources/${encodeURIComponent(name)}${params}`,
    `DELETE /sources/${name}`,
  );
}
